### PR TITLE
[ISSUE #450] fix connect fail when connect concurrently

### LINKED ISSUE #450: Some consumer trace send fail
1. Please describe the issue you observed:

- What did you do (The steps to reproduce)?
1. produce 3 messages
2. consume the messages with trace enabled
```C
int main(int argc, char* argv[]) {
  int i = 0;
  printf("PushConsumer Initializing....\n");
  CPushConsumer* consumer = CreatePushConsumer("Group_Consumer_Test");
  SetPushConsumerMessageTrace(consumer, OPEN);
  SetPushConsumerNameServerAddress(consumer, "127.0.0.1:9876");
  Subscribe(consumer, "TopicTest", "*");
  RegisterMessageCallback(consumer, doConsumeMessage);
  StartPushConsumer(consumer);
  printf("Push Consumer Start...\n");
  for (i = 0; i < 10; i++) {
    printf("Now Running : %d S\n", i * 10);
    thread_sleep(10000);
  }
  ShutdownPushConsumer(consumer);
  DestroyPushConsumer(consumer);
  printf("PushConsumer Shutdown!\n");
  return 0;
}
```
- What did you expect to see?
all the trace produce success

- What did you see instead?
some trace produce fail
some error is print in DefaultMQPushConsumerImpl::submitSendTraceRequest
but if I dont send some messages before starting consume, no error is print

2. Please tell us about your environment:

 - What is your OS?

 - What is your client version?

 - What is your RocketMQ version?

4. Other information (e.g. detailed explanation, logs, related issues, suggestions on how to fix, etc):

### PR DESCRIPTION
**Pull request by yuz10**
## What is the purpose of the change

fix #450

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when a cross-module dependency exists.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: src/transport/TcpRemotingClient.cpp [hunk 1]
```diff
@@ -319,8 +319,15 @@ std::shared_ptr<TcpTransport> TcpRemotingClient::CreateTransport(const string& a
         if (connectStatus == TCP_CONNECT_STATUS_SUCCESS) {
           return tcp;
         } else if (connectStatus == TCP_CONNECT_STATUS_WAIT) {
-          std::shared_ptr<TcpTransport> pTcp;
-          return pTcp;
+          connectStatus = tcp->waitTcpConnectEvent(static_cast<int>(m_tcpConnectTimeout));
```
Comment (ifplusor) [inline src/transport/TcpRemotingClient.cpp:322]: don't wait event during locking.

**Review by ifplusor — CHANGES_REQUESTED** [review]

File: src/transport/TcpRemotingClient.cpp (hunk 1 above)
Comment (yuz10) [reply src/transport/TcpRemotingClient.cpp:322]: Any suggustions? waitTcpConnectEvent is also called in the funtion if connection not in table

Comment (yuz10) [reply src/transport/TcpRemotingClient.cpp]: fixed.

File: src/transport/TcpRemotingClient.cpp [hunk 2]
```diff
@@ -334,8 +333,9 @@ std::shared_ptr<TcpTransport> TcpRemotingClient::CreateTransport(const string& a
 
     //<!callback;
     TcpTransportReadCallback callback = needResponse ? &TcpRemotingClient::static_messageReceived : nullptr;
-
-    tts = TcpTransport::CreateTransport(this, m_enableSsl, m_sslPropertyFile, callback);
+    if(!tts) {
+      tts = TcpTransport::CreateTransport(this, m_enableSsl, m_sslPropertyFile, callback);
+    }
     TcpConnectStatus connectStatus = tts->connect(addr, 0);  // use non-block
```
Comment (ifplusor) [inline src/transport/TcpRemotingClient.cpp:339]: If and only if create transport, call `connect`. watch change of status if status is wait.

**Review by ifplusor — CHANGES_REQUESTED** [review]

File: src/transport/TcpRemotingClient.cpp (hunk 2 above)
Comment (ifplusor) [reply src/transport/TcpRemotingClient.cpp:339]: and manage failure in the creating branch.

Comment (yuz10) [reply src/transport/TcpRemotingClient.cpp]: fixed

**Review by ifplusor — APPROVED** [review]

**Pull request merged**